bundles/org.platformio.eclipse.ide.home.ui/src/org/platformio/eclipse/ide/home/internal/ui/handlers/SelectProject.java

```diff
@@ -0,0 +1,91 @@
+/*******************************************************************************
+ * Copyright (c) 2021 PlatformIO and ArSysOp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Contributors:
+ *     Nikifor Fedorov (ArSysOp) - initial API and implementation
+ *******************************************************************************/
+package org.platformio.eclipse.ide.home.internal.ui.handlers;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.Path;
+import org.eclipse.jface.viewers.IStructuredSelection;
+import org.eclipse.jface.viewers.StructuredSelection;
+import org.eclipse.jface.window.Window;
+import org.eclipse.ui.PlatformUI;
+import org.eclipse.ui.dialogs.ElementListSelectionDialog;
+import org.eclipse.ui.model.WorkbenchLabelProvider;
+import org.platformio.eclipse.ide.home.core.Messages;
+
+public final class SelectProject implements Supplier<Optional<IProject>> {
+
+	private final IStructuredSelection selection;
+
+	public SelectProject(IStructuredSelection selection) {
+		this.selection = selection;
+	}
+
+	@Override
+	public Optional<IProject> get() {
+		Optional<IProject> selected = selected();
+		if (selected.isPresent()) {
+			return selected;
+		}
+		return ask();
+	}
+
+	private Optional<IProject> selected() {
+		if (!StructuredSelection.EMPTY.equals(selection)) {
+			Object element = selection.getFirstElement();
+			if (element instanceof IResource) {
+				IProject project = ((IResource) element).getProject();
+				if (project.exists(new Path("platformio.ini"))) {//$NON-NLS-1$
```

Comment:
you have a predicate, why not to start using it?
`selection.toList()`
and then stream (after filtering and ampping) could give you findFirst()